[Seah Bing Hang Benjamin] iP

sbhbenjamin opened the pull request:
# Duke ⚡️
> "Your mind is for having ideas, not holding them." - David Allen ([source](https://dansilvestre.com/productivity-quotes))

Duke frees your mind of having to remember things you need to do. Its:
- text-based
- easy to learn
- ~FAST~ *BLAZINGLY* FAST to use

All you need to do is:
1. Download it from [here](https://github.com/sbhbenjamin/ip/releases/tag/A-Jar).
2. Double-click it.
3. Add your tasks.
4. Let it manage for you! 🤗

And it is absolutely **FREE**!

Features:
- [x] Managing tasks
- [ ] Managing deadlines (coming soon)
- [ ] Reminders (coming soon)
---
If you are a Java programmer, you can use it to practice too. Here's the `main` method:
```java
public class Main {
        public static void main(String[] args) {
                Application.launch(MainApp.class, args);
        }
}
```

Commit:
Add Gradle support

sbhbenjamin committed:
Add level-1 features: greet, echo, exit

sbhbenjamin committed:
Add level-2 features: add, list

sbhbenjamin committed:
Complete level-3: mark as done and undone

sbhbenjamin committed:
Complete level-4: todos, events and deadlines

sbhbenjamin committed:
Complete Automated Text UI Testing

sbhbenjamin committed:
Complete level-5: handle errors

sbhbenjamin committed:
Complete level-6: delete

sbhbenjamin committed:
Level 6: Add tests for delete

sbhbenjamin committed:
Add command enum to validate commands

sbhbenjamin committed:
Add javadocs

sbhbenjamin committed:
Abstract out tasklist and add saving functionality

sbhbenjamin committed:
Add date formatting for event and deadline tasks

sbhbenjamin committed:
Merge branch 'branch-Level-7'

Merge branch branch-Level-7 into master

sbhbenjamin committed:
Resolve merge conflict (branch-Level-8 -> master)

sbhbenjamin committed:
Refactor code to better follow OOP conventions

sbhbenjamin committed:
Divide classes into packages

sbhbenjamin committed:
Merge branch 'add-gradle-support'

Merge branch 'add-gradle-support' into master

sbhbenjamin committed:
Add JUnit tests

sbhbenjamin committed:
Move command enum to correct directory

sbhbenjamin committed:
Add functionality to create file if not found and prepare for release

sbhbenjamin committed:
Add javadocs

sbhbenjamin committed:
Tweak code to comply to coding standard

sbhbenjamin committed:
Add task finding feature

sbhbenjamin committed:
Merge branch 'branch-A-CodingStandard'
Merge branch 'branch-A-CodingStandard' into master

sbhbenjamin committed:
Merge branch 'branch-Level-9'
Merge branch 'branch-Level-9' into master

sbhbenjamin committed:
Add checkstyle

sbhbenjamin committed:
Fix checkstyle errors

sbhbenjamin committed:
Add GUI

xSaints19x commented on src/main/java/duke/parser/Parser.java:
```diff
@@ -0,0 +1,59 @@
+        String command = input[0];
+        String arguments = input.length > 1 ? input[1] : "";
+
+        if (command.equals("bye")) {
+            return new ExitCommand();
+        } else if (command.equals("list")) {
+            return new InfoCommand();
+        } else if (command.equals("mark")) {
+            int index = Integer.parseInt(arguments);
+            return new UpdateCommand(true, index);
+        } else if (command.equals("unmark")) {
+            int index = Integer.parseInt(arguments);
+            return new UpdateCommand(false, index);
+        } else if (command.equals("todo")) {
+            return new AddCommand(arguments);
+        } else if (command.equals("deadline")) {
+            String[] argsArray = arguments.split(" /by ");
+            return new AddCommand(argsArray[0], argsArray[1], "DEADLINE");
+        } else if (command.equals("event")) {
+            String[] argsArray = arguments.split(" /at ");
+            return new AddCommand(argsArray[0], argsArray[1], "EVENT");
+        } else if (command.equals("find")) {
+            return new FindCommand(arguments);
+        } else if (command.equals("delete")) {
+            return new DeleteCommand(Integer.parseInt(arguments));
+        }
+        throw new InvalidCommandException("☹ OOPS!!! I'm sorry, but I don't know what that means :-(");
```
Perhaps a switch statement here would be good here?

xSaints19x reviewed (commented):
Overall looks good 👍
Just a small suggestion is given.

sbhbenjamin committed:
Update gitignore to include .idea/

AdrianOngJJ commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,28 @@
+            if (input.equals("bye")) {
+                break;
+            }
+            System.out.println("____________________________________________________________");
```
Instead of calling this command every time, may I suggest having a separate variable to store the line break and a separate method to print the line break

AdrianOngJJ commented on src/main/java/Duke.java:
```diff
@@ -67,6 +61,17 @@ public void listTasks() {
         System.out.println("____________________________________________________________");
     }
 
+    public void deleteTask(int index) {
```
Good job in abstracting it out

teominghuang commented on src/main/java/duke/commands/AddCommand.java:
```diff
@@ -52,13 +59,15 @@ public String execute() throws InvalidParameterException, IOException {
             }
             task = super.taskList.addDeadlineTask(description, deadline);
-            taskToString = String.format("%s|T|0|%s%n", task.getId(), task.getDescription(), ((DeadlineTask) task).getDeadline());
+            taskToString = String.format("%s|T|0|%s%n", task.getId(),
```
Really loved the format of the toString(), really intuitive and neat!

teominghuang commented on src/main/java/duke/parser/Parser.java:
```diff
@@ -1,6 +1,12 @@
 package duke.parser;
 
-import duke.commands.*;
```
Good job omitting the wildcard import and replacing it with very specific imports! :)

teominghuang commented on src/main/java/Storage.java:
```diff
@@ -0,0 +1,128 @@
+        List<String> fileContent = new ArrayList<>(Files.readAllLines(Paths.get("data/duke.txt"), StandardCharsets.UTF_8));
+        for (int i = 0; i < fileContent.size(); i++) {
+            String line = fileContent.get(i);
+            String[] a = line.split("\\|");
```
Variable names can be slightly more specific for collaborators to better collaborate with the code base. The structure of the code base is very good!

teominghuang commented on src/main/java/TaskList.java:
```diff
@@ -0,0 +1,161 @@
+        Task foundTask = taskList.get(index - 1);
+        taskList.remove(index - 1);
+        System.out.println("____________________________________________________________");
+        System.out.println("Noted. I've removed this task:");
```
My method was originally structured this way, however, I realized how it might render my system prone to bugs. Just a question, what would happen should the task fail to be removed from the tasklist? Will the system print out the removal of the task?

teominghuang reviewed (changes requested):
Overall, I thought your code base was decently structured. It is also interesting to see how you manage to utilise the string formats to deliver very concise formats. There is a slight tendency to use variables names that are not very intuitive. I have some questions pertaining to the logic flow for some of the methods. However all in all, very very good work! Loved it.

sbhbenjamin committed:
Modify build.gradle to generate javaFX JAR file

sbhbenjamin committed:
Add assertions for task and tasklist

Mostly added assertions to test post conditions, to ensure that the task list and task logic works correctly.

sbhbenjamin committed:
Refactor code quality

Some of the variables are not sufficiently semantic, and there exists many magic strings within the code.

The current code's readability can be improved.

Code quality should be improved to enhance understandability. Complicated expressions can be broken down into smaller ones, magic strings should be avoided, and variables should be better named to represent what it holds.

Let's break down complicated expressions into smaller ones, avoid magic strings, and better name variables to be more semantic.

Refer to this guide for more information about best practices to improve code quality:
https://nus-cs2103-ay2122s2.github.io/website/schedule/week5/topics.html#w5-4-code-quality

sbhbenjamin committed:
Merge pull request #1 from sbhbenjamin/branch-A-Assertions

Add assertions to Duke

sbhbenjamin committed:
Merge changes from assertions branch

Merge branch 'master' into branch-A-CodeQuality

sbhbenjamin committed:
Merge pull request #2 from sbhbenjamin/branch-A-CodeQuality

Improve code quality

sbhbenjamin committed:
Add CI

sbhbenjamin committed:
Fix style violations

sbhbenjamin committed:
Modify CI to exclude I/O redirection tests

sbhbenjamin committed:
Add tags

sbhbenjamin committed:
Fix checkstyle errors

sbhbenjamin committed:
Merge pull request #5 from sbhbenjamin/branch-C-Tagging

Add tags

sbhbenjamin committed:
Improve the GUI

sbhbenjamin committed:
Fix issue where tags are unable to be saved to memory

sbhbenjamin committed:
Prepare for final version

sbhbenjamin committed:
Add user guide

sbhbenjamin committed:
Fix event task test

sbhbenjamin committed:
Fix issue where table is not rendering with jekyll

sbhbenjamin committed:
Update user guide

sbhbenjamin committed:
Update readme to include jar link
